pkg/controllers/managementuser/rbac/roletemplates/roletemplate_handler.go

```diff
@@ -0,0 +1,232 @@
+package roletemplates
+
+import (
+	"errors"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	typesrbacv1 "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/wrangler/v3/pkg/slice"
+	v1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// TODO: use safeconcat for the names
+// TODO: handle external rules
+// TODO: Use a Wrangler client instead of Norman
+
+var promotedRulesForProjects = map[string]string{
+	"navlinks":          "ui.cattle.io",
+	"nodes":             "",
+	"persistentvolumes": "",
+	"storageclasses":    "storage.k8s.io",
+	"apiservices":       "apiregistration.k8s.io",
+	"clusterrepos":      "catalog.cattle.io",
+	"clusters":          "management.cattle.io",
+}
+
+const (
+	clusterRoleOwner = "authz.cluster.cattle.io/clusterrole-owner"
+)
+
+func newRoleTemplateLifecycle(uc *config.UserContext) *roleTemplateLifecycle {
+	return &roleTemplateLifecycle{
+		crClient: uc.RBAC.ClusterRoles(""),
+	}
+}
+
+type roleTemplateLifecycle struct {
+	crClient typesrbacv1.ClusterRoleInterface
+}
+
+// Create func always creates:
+// 1. a ClusterRole with the same name as the RoleTemplate
+// 2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates with the name "RoleTemplateName-aggregator"
+//
+// For RoleTemplates with the Context == "Project", we also create:
+// 1. If the RoleTemplate has any rules for Global Resources, make a ClusterRole with those named "RoleTemplateName-promoted"
+// 2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates' promoted Cluster Roles named "RoleTemplateName-promoted-aggregator"
+func (rtl *roleTemplateLifecycle) Create(rt *v3.RoleTemplate) (runtime.Object, error) {
+	// 1. Create cluster role
+	cr := rtl.buildClusterRole(rt)
+	_, err := rtl.crClient.Create(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 2. Create aggregating cluster role
+	cr = rtl.buildAggregatingClusterRole(rt)
+	_, err = rtl.crClient.Create(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 3. Project global resources
+	if rt.Context == "project" {
+		// Create a promoted cluster role
+		cr = rtl.buildProjectPromotedClusterRole(rt)
+		_, err = rtl.crClient.Create(cr)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create a promoted aggregating cluster role
+		cr = rtl.buildProjectPromotedAggregatedClusterRole(rt)
+		_, err = rtl.crClient.Create(cr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return rt, nil
+}
+
+func (rtl *roleTemplateLifecycle) Updated(rt *v3.RoleTemplate) (runtime.Object, error) {
+	// 1. Create cluster role
+	cr := rtl.buildClusterRole(rt)
+	_, err := rtl.crClient.Update(cr)
```

Comment:
This will fail if the resource does not exist (which could happen by accident?) entering in a failure loop.